sourced/ml/mining/utils/fs.py

```diff
@@ -13,8 +25,47 @@ def check_remove_filepath(path: Path, log: Logger, force: bool):
     """Checks the path does not point to a file, if it does will raise an error or remove it."""
     path.parent.mkdir(parents=True, exist_ok=True)
     if path.exists():
+        if not path.is_file():
+            log.error("%s exists and is not a file, aborting", path)
+            raise IsNotAFileError
         if not force:
             log.error("%s already exists, aborting (use -f/--force to overwrite)", path)
             raise FileExistsError
         log.warn("%s already exists, overwritten", path)
         path.unlink()
+
+
+def check_exists_filepath(path: Path, log: Logger):
+    """Checks the path points to a file, if it doesn't will raise an error."""
+    if not path.exists():
+        log.error("%s already exists, aborting", path)
+        raise FileNotFoundError
+
+
+def check_empty_directory(path: Path, log: Logger, force: bool, ignore: bool):
+    """Checks the path does not point to a non-empty directory, if it does will raise an error or
+    empty it."""
+    path.parent.mkdir(parents=True, exist_ok=True)
+    if path.exists():
+        if not path.is_dir():
+            log.error("%s exists and is not a directory, aborting", path)
+            raise IsNotADirectoryError
+        if not (force or ignore):
+            log.error(
+                "%s already exists, aborting (use -f/--force to empty or -i/--ignore to ignore)",
+                path,
+            )
+            raise NonEmptyDirectoryError
+        if ignore:
+            log.warn("%s already exists and is non-empty, ignoring anyway", path)
+            return
+        log.warn(
+            "%s already exists and is non-empty, removing all files and sub-directories",
+            path,
+        )
+        for p in sorted(path.glob("*/**"), reverse=True):
+            if p.is_dir():
+                p.rmdir()
+            else:
+                p.unlink()
```

Comment:
This does not work for directories

* https://docs.python.org/3/library/os.html#os.unlink
* https://docs.python.org/3/library/os.html#os.remove

You should use `shutil.rmtree` instead.

Reply:
Yeah but `p.rmdir` does, and since we are sorting the globbed paths in reverse the directories are emptied before they are removed. If you want I can modify the code to just use shutil, but this seemed more informative

Reply:
But https://docs.python.org/3/library/pathlib.html#pathlib.Path.rmdir says that the directory must be empty

Reply:
yeah, that's why we sort the filepaths in reverse order - so all the files aremode, since:
`dir/foo` will always be before `dir/`

Reply:
I see now. We should definitely use `shutil.rmtree` instead of the whole thing then

Reply:
okay np